pkg/cloudprovider/cloudprovider.go

```diff
@@ -1,62 +1,140 @@
 	failureDomainRegionKey = v1.LabelTopologyRegion
 )
 
 var (
+	instance             Ops
+	once                 sync.Once
 	providerRegistry     map[string]Ops
 	providerRegistryLock sync.Mutex
 )
 
 // Ops is a list of APIs to fetch information about cloudprovider and its nodes
 type Ops interface {
-	// Name returns the name of the cloud provider
+	// init checks environment and initializes cloud provider
+	init() error
+
+	// Name returns the providerName of the cloud provider
 	Name() string
 
+	// GetK8sDistributionName returns the k8s distribution providerName
+	GetK8sDistributionName() string
+
 	// GetZone returns the zone of the provided node
 	GetZone(*v1.Node) (string, error)
+
+	// CheckCloudDrivePermission checks if permissions for drive operation is granted
+	CheckCloudDrivePermission() error
 }
 
-// New returns a new implementation of the cloud provider
-func New(name string) Ops {
-	providerRegistryLock.Lock()
-	defer providerRegistryLock.Unlock()
+// Instance returns a singleton instance of cloud provider
+func Instance() Ops {
+	once.Do(func() {
+		if instance == nil {
+			instance = New("")
+		}
+	})
+	return instance
+}
 
-	ops, ok := providerRegistry[name]
-	if !ok {
-		return &defaultProvider{name}
+// SetInstance replaces the instance with the provided one. Should be used only for testing purposes.
+func SetInstance(i Ops) {
+	instance = i
+}
+
+// InitCloudProvider initialize the cloud provider instance at operator starts
+func InitCloudProvider() error {
+	initRegistry()
+
+	cloudProviderName, err := getCloudProviderName()
+	if err != nil {
+		return err
 	}
-	return ops
+
+	instance = New(cloudProviderName)
+	if err := instance.init(); err != nil {
+		return err
+	}
+
+	return nil
 }
 
-type defaultProvider struct {
-	name string
+func initRegistry() {
+	providerRegistryLock.Lock()
+	defer providerRegistryLock.Unlock()
+
+	providerRegistry = make(map[string]Ops)
+	providerRegistry[cloudops.Azure] = &azure{}
+	providerRegistry[string(cloudops.AWS)] = &aws{}
 }
 
-func (d *defaultProvider) Name() string {
-	return d.name
+func getCloudProviderName() (string, error) {
+	nodeList, err := coreops.Instance().GetNodes()
+	if err != nil {
+		return "", err
+	}
+
+	for _, node := range nodeList.Items {
+		// Get the cloud provider
+		// From kubernetes node spec: <ProviderName>://<ProviderSpecificNodeID>
+		if node.Spec.ProviderID != "" {
+			tokens := strings.Split(node.Spec.ProviderID, "://")
+			if len(tokens) == 2 {
+				return tokens[0], nil
+			} // else provider id is invalid
+		}
+	}
+	return "", nil
 }
 
-func (d *defaultProvider) GetZone(node *v1.Node) (string, error) {
-	if node == nil {
-		return "", fmt.Errorf("node cannot be nil")
+func getK8sDistributionName() (string, error) {
+	k8sVersion, err := coreops.Instance().GetVersion()
+	if err != nil {
+		return "", err
 	}
-	return node.Labels[failureDomainZoneKey], nil
+
+	logrus.WithFields(logrus.Fields{
+		"info": k8sVersion,
+	}).Infof("JLIAO: get k8s version")
```

Comment:
Remove before merge or make a "formal" log line?

Reply:
will address all log lines before merging, currently for easier debugging purpose.